rosidl_generator_java/resource/msg.cpp.em

```diff
@@ -90,31 +31,50 @@ list_jni_type = "java/util/List"
 array_list_normalized_type = "java__util__ArrayList"
 array_list_jni_type = "java/util/ArrayList"
 
+# Collect JNI types and includes
 cache = defaultdict(lambda: False)
-
 cache[msg_normalized_type] = msg_jni_type
-
-# We do not cache strings because java.lang.String behaves differently
-
-unique_fields = set()
-
-for field in spec.fields:
-    if field.type.is_array:
+namespaced_types = set()
+includes = set()
+for member in message.structure.members:
+    type_ = member.type
+    if isinstance(type_, AbstractNestedType):
         cache[list_normalized_type] = list_jni_type
         cache[array_list_normalized_type] = array_list_jni_type
-
-    if not field.type.type == 'string':
-        cache[get_normalized_type(field.type)] = get_jni_type(field.type)
-
-    if not field.type.is_primitive_type():
-        non_primitive_types.add(get_jni_type(field.type))
-        unique_fields.add((field.type.pkg_name, field.type.type))
+        type_ = type_.value_type
+        if isinstance(type_, BasicType):
+            includes.add('rosidl_generator_c/primitives_sequence.h')
+            includes.add('rosidl_generator_c/primitives_sequence_functions.h')
+
+    # We do not cache strings because java.lang.String behaves differently
+    if not isinstance(type_, AbstractGenericString):
+        cache[get_normalized_type(type_)] = get_jni_type(type_)
+
+    if isinstance(type_, AbstractString):
+        includes.add('rosidl_generator_c/string.h')
+        includes.add('rosidl_generator_c/string_functions.h')
+
+    if isinstance(type_, AbstractWString):
+        includes.add('rosidl_generator_c/u16string.h')
+        includes.add('rosidl_generator_c/u16string_functions.h')
+
+    if isinstance(type_, NamespacedType):
+        namespaced_types.add(get_jni_type(type_))
+        # TODO(jacobperron): double-check, we may have to handle service/action suffixes (see C generator)
```

Comment:
Doesn't seem like this is needed and the TODO can be removed.

Reply:
Could you remove this if it's not needed anymore? Thanks.

Reply:
Removed b59c5b2